### ISSUE TITLE
How to ensure message belonging to a single group is processed sequentially?

### ISSUE THREAD

**Issue by nripendra** [untu/comedy#19]
Here's what I'm trying to achieve:
- A cluster of actors to process commands (as in CQRS commands) 
- There will be 1 command router (that's where singleton question came from).
- All commands will go to command router actor.
- Each command is then distributed to command handler actors (from cluster mentioned in step 1).

My requirement is: command belonging to a single aggregate must be handled sequentially, although command belonging to different aggregates can be running in parallel.

Is it possible to setup the cluster of actors in such a way that when a command of an aggregate is being handled, other commands belonging to same aggregate will be queued until previous command is completed; while if command belonging to other aggregate comes in then it will be handled?

I'm envisioning solution like this:
- divide aggregate (using its Id) into N buckets, where N is number of actors in cluster. This way all commands belonging to an aggregate will be routed to only one actor, which will then hopefully be easy to process sequentially.

But this may have problem of single actor being more loaded if aggregateId belonging to bucket of that particular actor comes in frequently

**Comment by weekens** [comment]
@nripendra , you're correct. To achieve what you want, you'll need:
1. A handler actor instance for each aggregate type. These could be actors with different definitions if handling logic is different, or actors with the same definition if handling logic is the same.
2. A router actor that creates child actor instance for each aggregate during initialization (in `initialize()` method) and routes the command to the right handler actor depending on some routing logic that you will define in command handler method.

Something like this (just a scratch):

```javascript
var P = require('bluebird');

class CommandRouterActor {
  async initialize(selfActor) {
    this.handlersByCommandType = await P.reduce(handlerTypes, (memo, handlerCls, commandType) => {
      return selfActor.createChild(handlerCls).then(actor => {
        memo[commandType] = actor;

        return memo;
      });
    }, {});
  }

  handleCommand(command) {
    var handlerActor = this.handlersByCommandType[command.type];

    if (!handlerActor) throw new Error('Unknown command type:' + command.type);

    return handlerActor.sendAndReceive('handle', command);
  }
}
```

About the load distribution: you can only distribute load by executing commands in parallel (creating more instances of a command handler actor of a particular command type in `"forked"` or `"remote"` mode). If sequential command processing for a single command type is a hard requirement, you'll have to live with a single instance of a command handler actor for each command type.

**Comment by nripendra** [comment]
@weekens My actors will be able to handle any type of commands, so I don't need division of actor by comandType. There is further division of handler functions where command needs to be delivered by these worker actors. Each actor can be forked or remote but command belonging to an aggregate should be handled by same actor. I can change actor to further fork itself if necessary, such a way that it again uniformly distributes commands to it's children following similar rule.

Here's rough modification of code that I'm thinking. Will this kind of setup work?
```js
var P = require('bluebird');

class CommandRouterActor {
  async initialize(selfActor) {
    this.workers = [];
    for(let i = 0; i< 4 ; i++) {
       this.workers.push(await selfActor.createChild('./worker'));
    }
  }

  handleCommand(command) {
    const hash = hashCode(command.aggregateId);
      let workerNumber = Math.floor(
        hash % 4 // 4 worker actors
      );
    var handlerActor = this.workers[workerNumber];


    // if (!handlerActor) throw new Error('Unknown command type:' + command.type);

    return handlerActor.sendAndReceive('handle', command);
  }
}

// convert string to some integer
// There is no real necessity for integer to be hard unique.
function hashCode(str) {
  var hash = 0,
    i,
    chr;
  if (str.length === 0) return hash;
  for (i = 0; i < str.length; i++) {
    chr = str.charCodeAt(i);
    hash = (hash << 5) - hash + chr;
    hash |= 0;
  }
  return hash;
}
```

**Comment by weekens** [comment]
@nripendra , yeah, that looks good. But you better make `hashCode()` a private method of `CommandRouterActor`. This way `CommandHandlerActor` definition will be completely serializable, which is good for Comedy.
